File: rem/commands/registry.py
"""Command registry: dataclass + REGISTRY list + handlers for all slash commands."""

from __future__ import annotations

import json
import subprocess
import sys
import termios
import time
import tty
from dataclasses import dataclass, field
from pathlib import Path
from typing import Callable, Optional

from rich import box
from rich.console import Group
from rich.panel import Panel
from rich.table import Table

from rem import config
from rem.ui.theme import (
    get_active_theme,
    get_console,
    list_theme_names,
    set_active_theme,
)


@dataclass(frozen=True)
class Command:
    """A single slash command entry."""

    name: str
    description: str
    shortcut: str
    handler: Callable[[], None]


THEME_DESCRIPTIONS: dict[str, str] = {
    "GHOST": "dark monochrome",
    "PHOSPHOR": "green terminal",
    "MIST": "blue calm",
    "EMBER": "warm orange",
    "SAKURA": "pink/purple",
    "PAPER": "light beige",
}


conversation_history: list[dict[str, str]] = []


def _read_raw_key() -> Optional[str]:
    """Read a single keypress from stdin in raw mode; None for ESC."""
    fd = sys.stdin.fileno()
    old_settings = termios.tcgetattr(fd)
    try:
        tty.setraw(fd)
        ch = sys.stdin.read(1)
    finally:
        termios.tcsetattr(fd, termios.TCSADRAIN, old_settings)
    if ch == "\x1b":
        return None
    if ch in ("\r", "\n"):
        return None
    return ch


def _theme_picker_panel() -> Panel:
    """Render the numbered theme picker panel."""
    t = get_active_theme()
    console = get_console()
    lines = ["[bold]Pick a theme:[/]\n"]
    names = list_theme_names()
    for i, name in enumerate(names, start=1):
        desc = THEME_DESCRIPTIONS.get(name, "")
        lines.append(
            f"  [{t.accent}]{i}.[/] [{t.accent_dim}]{name:<9}[/]  [{t.text_muted}]{desc}[/]"
        )
    lines.append(f"\n  [{t.text_muted}]Press 1-6 to pick, ESC to cancel.[/]")
    body = "\n".join(lines)
    return Panel(
        body,
        border_style=t.border,
        box=box.HORIZONTALS,
        padding=(0, 1),
        title=f"[{t.accent}]THEMES[/]",
        title_align="left",
    )


def _model_picker_panel(models: list[str], current: str) -> Panel:
    """Render the numbered model picker panel."""
    t = get_active_theme()
    lines = ["[bold]Available models:[/]\n"]
    if not models:
        lines.append(f"  [{t.text_muted}](no models found via `ollama list`)[/]")
    for i, name in enumerate(models, start=1):
        marker = "*" if name == current else " "
        lines.append(
            f"  [{t.accent}]{i}.[/] [{t.accent_dim}]{name}[/]  [{t.text_muted}]{marker}[/]"
        )
    lines.append(
        f"\n  [{t.text_muted}]Enter a number and press Enter, or just Enter to cancel.[/]"
    )
    body = "\n".join(lines)
    return Panel(
        body,
        border_style=t.border,
        box=box.HORIZONTALS,
        padding=(0, 1),
        title=f"[{t.accent}]MODELS[/]",
        title_align="left",
    )


def _fetch_ollama_models() -> list[str]:
    """Run `ollama list` and parse the first column; empty list on failure."""
    try:
        result = subprocess.run(
            ["ollama", "list"],
            capture_output=True,
            text=True,
            timeout=5,
            check=False,
        )
    except (FileNotFoundError, subprocess.TimeoutExpired, OSError):
        return []
    if result.returncode != 0:
        return []
    models: list[str] = []
    for line in result.stdout.splitlines():
        line = line.strip()
        if not line:
            continue
        upper = line.upper()
        if (
            upper.startswith("NAME")
            or upper.startswith("NAME\t")
            or upper.startswith("NAME ")
        ):
            continue
        first = line.split()[0]
        if first:
            models.append(first)
    return models


def _cmd_help() -> None:
    """Render a Rich table of all registered commands."""
    t = get_active_theme()
    console = get_console()
    table = Table(
        box=box.SIMPLE,
        show_header=True,
        header_style=f"bold {t.accent}",
        border_style=t.border,
        expand=False,
    )
    table.add_column("name", style=t.accent, no_wrap=True)
    table.add_column("description", style=t.text_muted)
    table.add_column("shortcut", style=t.kbd_text, justify="right")
    for cmd in REGISTRY:
        table.add_row(cmd.name, cmd.description, cmd.shortcut or "")
    console.print(table)


def _cmd_mode() -> None:
    """Toggle CHAT <-> CODE in config and re-render the header."""
    t = get_active_theme()
    console = get_console()
    current = str(config.get("mode") or "CHAT")
    new_mode = "CODE" if current == "CHAT" else "CHAT"
    config.set_value("mode", new_mode)
    console.print(f"[{t.accent}]Mode switched to {new_mode}.[/]")
    from rem.ui.header import render_header

    render_header(str(config.get("model") or "rem-coder"), new_mode)


def _cmd_theme() -> None:
    """Clear screen, show picker, read 1-6 or ESC, apply choice, re-render header."""
    t = get_active_theme()
    console = get_console()
    console.clear()
    console.print(_theme_picker_panel())
    key = _read_raw_key()
    if key is None:
        return
    if not key.isdigit():
        console.print(f"[{t.text_muted}]Cancelled.[/]")
        return
    idx = int(key) - 1
    names = list_theme_names()
    if idx < 0 or idx >= len(names):
        console.print(f"[{t.text_muted}]Cancelled.[/]")
        return
    chosen = names[idx]
    set_active_theme(chosen)
    t = get_active_theme()
    console.print(f"[{t.accent}]Theme switched to {chosen}.[/]")
    from rem.ui.header import render_header

    render_header(
        str(config.get("model") or "rem-coder"), str(config.get("mode") or "CHAT")
    )


def _cmd_model() -> None:
    """Fetch ollama list, show picker, save the chosen model to config."""
    t = get_active_theme()
    console = get_console()
    models = _fetch_ollama_models()
    current = str(config.get("model") or "")
    console.print(_model_picker_panel(models, current))
    try:
        raw = input().strip()
    except EOFError:
        return
    if not raw:
        console.print(f"[{t.text_muted}]Cancelled.[/]")
        return
    if not raw.isdigit():
        console.print(f"[{t.text_muted}]Invalid selection.[/]")
        return
    idx = int(raw) - 1
    if idx < 0 or idx >= len(models):
        console.print(f"[{t.text_muted}]Invalid selection.[/]")
        return
    chosen = models[idx]
    config.set_value("model", chosen)
    console.print(f"[{t.accent}]Model set to {chosen}.[/]")
    from rem.ui.header import render_header

    render_header(chosen, str(config.get("mode") or "CHAT"))


def _cmd_clear() -> None:
    """Empty the conversation history list."""
    t = get_active_theme()
    console = get_console()
    conversation_history.clear()
    console.print(f"[{t.accent}]History cleared.[/]")


def _cmd_save() -> None:
    """Persist conversation_history to rem_session_{timestamp}.json in cwd."""
    t = get_active_theme()
    console = get_console()
    ts = int(time.time())
    path = Path(f"rem_session_{ts}.json")
    payload = {
        "timestamp": ts,
        "history": list(conversation_history),
        "config": config.load_config(),
    }
    with path.open("w", encoding="utf-8") as handle:
        json.dump(payload, handle, indent=2)
    console.print(f"[{t.accent}]Saved to {path}.[/]")


def _cmd_exit() -> None:
    """Print goodbye and terminate the process."""
    t = get_active_theme()
    console = get_console()
    console.print(f"[{t.sys_color}]Goodbye.[/]")
    sys.exit(0)


REGISTRY: list[Command] = [
    Command(
        name="/help",
        description="show this help table",
        shortcut="?",
        handler=_cmd_help,
    ),
    Command(
        name="/mode",
        description="toggle CHAT <-> CODE",
        shortcut="m",
        handler=_cmd_mode,
    ),
    Command(
        name="/model",
        description="switch the active model",
        shortcut="M",
        handler=_cmd_model,
    ),
    Command(
        name="/theme",
        description="switch color theme",
        shortcut="t",
        handler=_cmd_theme,
    ),
    Command(
        name="/clear",
        description="clear conversation history",
        shortcut="c",
        handler=_cmd_clear,
    ),
    Command(
        name="/save",
        description="save session to JSON",
        shortcut="s",
        handler=_cmd_save,
    ),
    Command(name="/exit", description="quit rem", shortcut="q", handler=_cmd_exit),
]


def find_command(name: str) -> Optional[Command]:
    """Look up a Command by exact name in REGISTRY."""
    for cmd in REGISTRY:
        if cmd.name == name:
            return cmd
    return None

File: rem-cli/src/commands/registry.rs
// ── commands/registry.rs ── cli was written in rust so dont write in py same design but redesign the current cli
//
// Slash command registry: every `/...` command that the REPL recognizes.
// Handlers are plain functions; state is held in module-level `RefCell`
// only for the conversation history. The header, theme, and config are
// always re-read fresh from disk inside each handler so live edits from
// other processes are picked up.
#![allow(dead_code)]

use std::cell::RefCell;
use std::io::{self, Write};
use std::process::Command as StdCommand;
use std::time::{SystemTime, UNIX_EPOCH};

use crate::config;
use crate::ui::theme;

#[derive(Debug, Clone)]
pub struct Command {
    pub name: &'static str,
    pub description: &'static str,
    pub shortcut: &'static str,
    pub handler: fn(),
}

pub const REGISTRY: &[Command] = &[
    Command {
        name: "/help",
        description: "show all commands and keybindings",
        shortcut: "?",
        handler: cmd_help,
    },
    Command {
        name: "/mode",
        description: "switch CHAT <-> CODE",
        shortcut: "m",
        handler: cmd_mode,
    },
    Command {
        name: "/model",
        description: "change active model",
        shortcut: "M",
        handler: cmd_model,
    },
    Command {
        name: "/theme",
        description: "change color theme",
        shortcut: "t",
        handler: cmd_theme,
    },
    Command {
        name: "/clear",
        description: "clear conversation history",
        shortcut: "c",
        handler: cmd_clear,
    },
    Command {
        name: "/save",
        description: "save session to file",
        shortcut: "s",
        handler: cmd_save,
    },
    Command {
        name: "/exit",
        description: "quit rem",
        shortcut: "q",
        handler: cmd_exit,
    },
];

/// Lookup by exact name (e.g. "/help"). Returns `None` for unknown.
pub fn find(name: &str) -> Option<&'static Command> {
    REGISTRY.iter().find(|c| c.name == name)
}

thread_local! {
    static HISTORY: RefCell<Vec<(String, String)>> = RefCell::new(Vec::new());
}

pub fn history() -> Vec<(String, String)> {
    HISTORY.with(|h| h.borrow().clone())
}

pub fn push_history(user: String, assistant: String) {
    HISTORY.with(|h| h.borrow_mut().push((user, assistant)));
}

pub fn clear_history() {
    HISTORY.with(|h| h.borrow_mut().clear());
}

// ── Handlers ──────────────────────────────────────────────────────────────

fn cmd_help() {
    let t = theme::active();
    let border = theme::paint(&t.border, "\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}\u{2500}", false);
    theme::println(&format!(
        "  {} {}",
        theme::paint(&t.accent, "COMMANDS", true),
        border
    ));

    let name_w = REGISTRY.iter().map(|c| c.name.len()).max().unwrap_or(8);
    let desc_w = REGISTRY
        .iter()
        .map(|c| c.description.len())
        .max()
        .unwrap_or(20);
    for cmd in REGISTRY.iter() {
        let name = theme::paint(
            &t.accent,
            &format!("  {:<width$}", cmd.name, width = name_w),
            true,
        );
        let desc = theme::paint(
            &t.text_muted,
            &format!("{:<width$}", cmd.description, width = desc_w),
            false,
        );
        let sc = if cmd.shortcut.is_empty() {
            "".to_string()
        } else {
            theme::paint_on(
                &t.kbd_text,
                &t.kbd_bg,
                &format!(" {} ", cmd.shortcut),
                false,
            )
        };
        theme::println(&format!("{name}  {desc}  {sc}"));
    }
    theme::println("");
    theme::println(&format!(
        "  {}",
        theme::paint(
            &t.text_faint,
            "tip: press / on an empty prompt to open the palette",
            false
        )
    ));
}

fn cmd_mode() {
    let t = theme::active();
    let mut cfg = config::load_config();
    cfg.mode = if cfg.mode == "CHAT" {
        "CODE".to_string()
    } else {
        "CHAT".to_string()
    };
    let _ = config::save_config(&cfg);
    theme::println(&format!(
        "  {} {}",
        theme::paint(&t.accent, "▌ mode switched to", true),
        theme::paint(&t.accent, &cfg.mode, true),
    ));
    crate::ui::header::render(&cfg.model, &cfg.mode);
}

fn cmd_theme() {
    let names = theme::list_names();
    render_theme_picker(&names);
    match theme_picker_read(&names) {
        Some(idx) => {
            let chosen = names[idx].clone();
            if theme::set_active(&chosen) {
                let t = theme::active();
                theme::println(&format!(
                    "  {} {}",
                    theme::paint(&t.accent, "▌ theme switched to", true),
                    theme::paint(&t.accent, &chosen, true),
                ));
                let cfg = config::load_config();
                crate::ui::header::render(&cfg.model, &cfg.mode);
            }
        }
        None => {
            let t = theme::active();
            theme::println(&format!(
                "  {}",
                theme::paint(&t.text_muted, "Cancelled.", false)
            ));
        }
    }
}

fn cmd_model() {
    let models = fetch_ollama_models();
    if models.is_empty() {
        let t = theme::active();
        theme::println(&format!(
            "  {}",
            theme::paint(&t.text_muted, "(no models found via `ollama list`)", false)
        ));
        return;
    }
    render_model_picker(&models);
    match theme_picker_read(&models) {
        Some(idx) => {
            let mut cfg = config::load_config();
            cfg.model = models[idx].clone();
            let _ = config::save_config(&cfg);
            let t = theme::active();
            theme::println(&format!(
                "  {} {}",
                theme::paint(&t.accent, "▌ model set to", true),
                theme::paint(&t.accent, &cfg.model, true),
            ));
            crate::ui::header::render(&cfg.model, &cfg.mode);
        }
        None => {
            let t = theme::active();
            theme::println(&format!(
                "  {}",
                theme::paint(&t.text_muted, "Cancelled.", false)
            ));
        }
    }
}

fn cmd_clear() {
    clear_history();
    let t = theme::active();
    theme::println(&format!(
        "  {}",
        theme::paint(&t.accent, "History cleared.", true)
    ));
}

fn cmd_save() {
    let t = theme::active();
    let ts = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs())
        .unwrap_or(0);
    let path = format!("rem_session_{ts}.json");
    let payload = serde_json::json!({
        "timestamp": ts,
        "history": history(),
        "config": config::load_config(),
    });
    let serialized = match serde_json::to_string_pretty(&payload) {
        Ok(s) => s,
        Err(e) => {
            theme::println(&format!(
                "  {}",
                theme::paint(&t.sys_color, &format!("save failed: {e}"), false)
            ));
            return;
        }
    };
    match std::fs::write(&path, serialized) {
        Ok(()) => theme::println(&format!(
            "  {} {}",
            theme::paint(&t.accent, "▌ saved to", true),
            theme::paint(&t.accent_dim, &path, false),
        )),
        Err(e) => theme::println(&format!(
            "  {}",
            theme::paint(&t.sys_color, &format!("save failed: {e}"), false)
        )),
    }
}

fn cmd_exit() {
    let t = theme::active();
    theme::println(&format!(
        "  {}",
        theme::paint(&t.text_muted, "Goodbye.", false)
    ));
    std::process::exit(0);
}

// ── Pickers (theme + model) ───────────────────────────────────────────────

fn render_theme_picker(names: &[String]) {
    let t = theme::active();
    let cfg = config::load_config();
    theme::println(&format!(
        "  {}",
        theme::paint(&t.accent, "SELECT THEME", true)
    ));
    let w = names.iter().map(|s| s.len()).max().unwrap_or(6);
    for (i, name) in names.iter().enumerate() {
        let marker = if name.eq_ignore_ascii_case(&cfg.theme) {
            "●"
        } else {
            " "
        };
        let num = format!("  {} ", i + 1);
        let label = theme::paint(&t.accent, &num, true);
        let key = theme::paint(&t.accent, &format!("{:<w$}", name, w = w), true);
        let mut desc = String::new();
        desc.push_str(theme_short_desc(name));
        let desc_colored = theme::paint(&t.text_muted, &desc, false);
        let marker_colored = theme::paint(&t.accent_dim, marker, false);
        theme::println(&format!("{label} {key}  {desc_colored}  {marker_colored}"));
        // Live preview line
        crate::ui::header::render_with_theme(&cfg.model, &cfg.mode, name);
    }
    theme::println("");
    theme::println(&format!(
        "  {}",
        theme::paint(&t.text_faint, "press 1-6 to switch, ESC to cancel", false)
    ));
    let _ = io::stdout().flush();
}

fn theme_short_desc(name: &str) -> &'static str {
    match name {
        "GHOST" => "pure black · ghost white accents",
        "PHOSPHOR" => "CRT green · retro terminal",
        "MIST" => "slate gray · cool blue tones",
        "EMBER" => "warm amber · charcoal dark",
        "SAKURA" => "deep navy · pink accents",
        "PAPER" => "off-white · ink on vellum",
        _ => "",
    }
}

fn render_model_picker(models: &[String]) {
    let t = theme::active();
    let cfg = config::load_config();
    theme::println(&format!(
        "  {}",
        theme::paint(&t.accent, "SELECT MODEL", true)
    ));
    let w = models.iter().map(|s| s.len()).max().unwrap_or(6);
    for (i, name) in models.iter().enumerate() {
        let marker = if name == &cfg.model { "●" } else { " " };
        let num = format!("  {} ", i + 1);
        let label = theme::paint(&t.accent, &num, true);
        let key = theme::paint(&t.accent_dim, &format!("{:<w$}", name, w = w), false);
        let marker_colored = theme::paint(&t.accent, marker, true);
        theme::println(&format!("{label} {key}  {marker_colored}"));
    }
    theme::println("");
    theme::println(&format!(
        "  {}",
        theme::paint(&t.text_faint, "press a number, or ESC to cancel", false)
    ));
    let _ = io::stdout().flush();
}

fn theme_picker_read(options: &[String]) -> Option<usize> {
    use std::io::Read;
    let stdin = io::stdin();
    let mut handle = stdin.lock();
    let mut buf = [0u8; 1];
    let mut collected = String::new();
    loop {
        let n = match handle.read(&mut buf) {
            Ok(n) => n,
            Err(_) => return None,
        };
        if n == 0 {
            return None;
        }
        match buf[0] {
            0x1b => return None,
            b'\r' | b'\n' => {
                if let Ok(idx) = collected.trim().parse::<usize>() {
                    if idx >= 1 && idx <= options.len() {
                        return Some(idx - 1);
                    }
                }
                return None;
            }
            0x7f | 0x08 => {
                collected.pop();
            }
            c if c.is_ascii_digit() => collected.push(c as char),
            _ => {}
        }
    }
}

fn fetch_ollama_models() -> Vec<String> {
    let output = StdCommand::new("ollama").arg("list").output();
    let output = match output {
        Ok(o) => o,
        Err(_) => return Vec::new(),
    };
    if !output.status.success() {
        return Vec::new();
    }
    let text = match std::str::from_utf8(&output.stdout) {
        Ok(s) => s,
        Err(_) => return Vec::new(),
    };
    let mut out = Vec::new();
    for line in text.lines() {
        let line = line.trim();
        if line.is_empty() {
            continue;
        }
        let upper = line.to_ascii_uppercase();
        if upper.starts_with("NAME") {
            continue;
        }
        let first = line.split_whitespace().next().unwrap_or("");
        if !first.is_empty() {
            out.push(first.to_string());
        }
    }
    out
}

/// Render a themed "section header" used by subcommands (`new`, `ask`,
/// `explain`, `patch`). Two lines: a bold title bar in the active accent,
/// then a faint subtitle line.
pub fn print_section(title: &str, subtitle: &str) {
    let t = theme::active();
    let bar = "\u{2500}".repeat(48usize.saturating_sub(title.chars().count()));
    theme::println(&format!(
        "  {} {}",
        theme::paint(&t.accent, title, true),
        theme::paint(&t.border, &bar, false),
    ));
    if !subtitle.is_empty() {
        theme::println(&format!(
            "  {}",
            theme::paint(&t.text_muted, subtitle, false),
        ));
    }
}

/// Print a themed success message. Equivalent to the old `style!(C_GREEN, "✓")` line.
pub fn print_success(msg: &str) {
    let t = theme::active();
    theme::println(&format!(
        "  {} {}",
        theme::paint(&t.accent, "\u{2713}", true),
        theme::paint(&t.accent, msg, true),
    ));
}

/// Print a themed file/entry row (icon + path + size).
pub fn print_file_row(icon: &str, path: &str, size: usize) {
    let t = theme::active();
    theme::println(&format!(
        "    {} {} {}",
        theme::paint(&t.text_faint, icon, false),
        theme::paint(&t.accent, path, false),
        theme::paint(&t.text_muted, &format!("({} bytes)", size), false),
    ));
}

/// Print a faint hint line at the end of an operation.
pub fn print_hint(msg: &str) {
    let t = theme::active();
    theme::println(&format!(
        "  {} {}",
        theme::paint(&t.text_faint, "next:", true),
        theme::paint(&t.text_muted, msg, false),
    ));
}

/// Theme-colored file icon for a given path.
pub fn file_icon_for(path: &str) -> &'static str {
    let lower = path.to_ascii_lowercase();
    if lower.ends_with(".html") || lower.ends_with(".htm") {
        "\u{1F310}"
    } else if lower.ends_with(".css") {
        "\u{1F3A8}"
    } else if lower.ends_with(".js") || lower.ends_with(".mjs") || lower.ends_with(".ts") {
        "\u{26A1}"
    } else if lower.ends_with(".json") {
        "\u{1F4CB}"
    } else if lower.ends_with(".md") || lower.ends_with(".txt") {
        "\u{1F4C4}"
    } else if lower.ends_with(".py") {
        "\u{1F40D}"
    } else {
        "\u{1F4C4}"
    }
}
